File: iosApp/iosApp/ui/payment/SettlementAccountViewModel.swift
import Combine
import Foundation
import Shared

// Compose SettlementAccountViewModel.kt 미러 — 정산 계좌 조회·저장 (M-12-3)
final class SettlementAccountViewModel: ObservableObject {
    private let getSettlementAccountUseCase: GetSettlementAccountUseCase

    private let saveSettlementAccountUseCase: SaveSettlementAccountUseCase

    @Published private(set) var uiState = SettlementAccountUiState()

    let event = PassthroughSubject<SettlementAccountEvent, Never>()

    private var hasEntered = false

    private func onEnter() {
        if hasEntered {
            return
        }
        hasEntered = true
        load()
    }

    private func load() {
        uiState.isLoading = true
        getSettlementAccountUseCase.invoke { [weak self] result, error in
            DispatchQueue.main.async {
                guard let self else { return }
                let account = (result as? AppResultSuccess<AnyObject>)?.value as? SettlementAccount

                self.uiState.isLoading = false
                if error == nil, let account {
                    self.uiState.bankName = account.bankName
                    // 조회는 마스킹된 번호만 준다 — 그대로 저장하면 실제 번호가 덮인다.
                    // 편집 필드는 비우고 마스킹 값은 안내로만 보여준다.
                    self.uiState.accountNumber = ""
                    self.uiState.maskedAccountNumber = account.maskedAccountNumber
                    self.uiState.holderName = account.holderName
                } else if let appError = (result as? AppResultFailure)?.error, !(appError is AppError.NotFound) {
                    // 미등록(404)은 빈 폼으로 시작한다 (M-12-3)
                    self.event.send(.showNotice(message: "계좌 정보를 불러오지 못했어요"))
                }
            }
        }
    }

    private func onSubmit() {
        let state = uiState

        if !state.canSubmit {
            return
        }
        uiState.isSubmitting = true

        let account = SettlementAccount(
            bankName: state.bankName,
            maskedAccountNumber: state.accountNumber,
            holderName: state.holderName
        )

        saveSettlementAccountUseCase.invoke(account: account) { [weak self] result, error in
            DispatchQueue.main.async {
                guard let self else { return }
                self.uiState.isSubmitting = false
                if error == nil, result is AppResultSuccess<AnyObject> {
                    self.event.send(.saved)
                } else {
                    let appError = (result as? AppResultFailure)?.error

                    self.event.send(.showNotice(message: self.saveFailMessage(appError)))
                }
            }
        }
    }

    // 서버 code 기반 문구 분기 (규칙 §10)
    private func saveFailMessage(_ error: AppError?) -> String {
        if let validation = error as? AppError.ValidationFailed {
            return validation.serverMessage ?? "계좌 정보를 확인해 주세요"
        } else if error is AppError.NetworkError {
            return "네트워크 연결을 확인해 주세요"
        } else {
            return "계좌를 저장하지 못했어요. 다시 시도해 주세요"
        }
    }

    func action(_ action: SettlementAccountAction) {
        switch action {
        case .enter:
            onEnter()
        case let .changeBankName(text):
            uiState.bankName = text
        case let .changeAccountNumber(text):
            uiState.accountNumber = String(text.filter { $0.isNumber || $0 == "-" }.prefix(20))
        case let .changeHolderName(text):
            uiState.holderName = text
        case .submit:
            onSubmit()
        }
    }

    init(
        getSettlementAccountUseCase: GetSettlementAccountUseCase,
        saveSettlementAccountUseCase: SaveSettlementAccountUseCase
    ) {
        self.getSettlementAccountUseCase = getSettlementAccountUseCase
        self.saveSettlementAccountUseCase = saveSettlementAccountUseCase
    }
}

File: composeApp/src/commonMain/kotlin/org/sesacteamproject/passmate/ui/payment/SettlementAccountViewModel.kt
package org.sesacteamproject.passmate.ui.payment

import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import org.sesacteamproject.passmate.core.model.AppError
import org.sesacteamproject.passmate.core.model.onFailure
import org.sesacteamproject.passmate.core.model.onSuccess
import org.sesacteamproject.passmate.mvi.MviViewModel
import org.sesacteamproject.passmate.payment.domain.model.SettlementAccount
import org.sesacteamproject.passmate.payment.domain.usecase.GetSettlementAccountUseCase
import org.sesacteamproject.passmate.payment.domain.usecase.SaveSettlementAccountUseCase

class SettlementAccountViewModel(
    private val getSettlementAccountUseCase: GetSettlementAccountUseCase,
    private val saveSettlementAccountUseCase: SaveSettlementAccountUseCase
) : MviViewModel<SettlementAccountUiState, SettlementAccountAction, SettlementAccountEvent>(SettlementAccountUiState()) {

    private var hasEntered = false

    private fun onEnter() {
        if (hasEntered) {
            return
        }
        hasEntered = true
        load()
    }

    private fun load() {
        _uiState.update { it.copy(isLoading = true) }
        viewModelScope.launch {
            getSettlementAccountUseCase.invoke()
                .onSuccess { account ->
                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            bankName = account.bankName,
                            // 조회는 마스킹된 번호만 준다 — 그대로 저장하면 실제 번호가 덮인다.
                            // 편집 필드는 비우고 마스킹 값은 안내로만 보여준다.
                            accountNumber = "",
                            maskedAccountNumber = account.maskedAccountNumber,
                            holderName = account.holderName
                        )
                    }
                }
                .onFailure { error ->
                    // 미등록(404)은 빈 폼으로 시작한다 (M-12-3)
                    _uiState.update { it.copy(isLoading = false) }
                    if (error !is AppError.NotFound) {
                        _event.emit(SettlementAccountEvent.ShowNotice("계좌 정보를 불러오지 못했어요"))
                    }
                }
        }
    }

    private fun onSubmit() {
        val state = _uiState.value

        if (!state.canSubmit) {
            return
        }
        _uiState.update { it.copy(isSubmitting = true) }
        viewModelScope.launch {
            val account = SettlementAccount(
                bankName = state.bankName,
                maskedAccountNumber = state.accountNumber,
                holderName = state.holderName
            )

            saveSettlementAccountUseCase.invoke(account)
                .onSuccess {
                    _uiState.update { it.copy(isSubmitting = false) }
                    _event.emit(SettlementAccountEvent.Saved)
                }
                .onFailure { error ->
                    _uiState.update { it.copy(isSubmitting = false) }
                    _event.emit(SettlementAccountEvent.ShowNotice(saveFailMessage(error)))
                }
        }
    }

    // 서버 code 기반 문구 분기 (규칙 §10)
    private fun saveFailMessage(error: AppError): String {
        return if (error is AppError.ValidationFailed) {
            error.serverMessage ?: "계좌 정보를 확인해 주세요"
        } else if (error is AppError.NetworkError) {
            "네트워크 연결을 확인해 주세요"
        } else {
            "계좌를 저장하지 못했어요. 다시 시도해 주세요"
        }
    }

    override fun onAction(action: SettlementAccountAction) {
        when (action) {
            is SettlementAccountAction.Enter -> onEnter()
            is SettlementAccountAction.ChangeBankName -> _uiState.update { it.copy(bankName = action.text) }
            is SettlementAccountAction.ChangeAccountNumber -> _uiState.update {
                it.copy(accountNumber = action.text.filter { ch -> ch.isDigit() || ch == '-' }.take(20))
            }
            is SettlementAccountAction.ChangeHolderName -> _uiState.update { it.copy(holderName = action.text) }
            is SettlementAccountAction.Submit -> onSubmit()
        }
    }
}
